### PR TITLE
Use predefined sets of naming styles instead of explicit regexps

### PR DESCRIPTION
**Pull request by rogalski**
An idea mentioned in issue #1013. 

I am aware it's definitely not ready for merge, but I'd leave it here just to allow idea to sink and to hear feedback from you guys. 

There are additional steps required for this change to be possibly merged:
- [x] Implement functionality of predefined presets (while keeping possibility of custom overrides by `something_rgx` options)
- [x] Select coding style presets to be included (possible candidates: snake_case (default), camelCase, PascalCase)
- [x] Define regular expressions for selected presets (I did not paid much attention to it in `CamelCaseStyle`)
- [x] Extend test suite for enough coverage

It may be a good idea to run a change on some big code bases with selected coding styles just to make sure we have covered all corner cases. We should aim for no false-positives (even if some false-negatives occurs), especially given that Pylint 2.0 is supposed to be a lot less nitpicking.

### CODE REVIEW

**Comment by rogalski** [comment]
Travis fails are two misaligned brackets from pylint report and pypy fail, which passes on my local machine. Likely it's an environment issue.

**Comment by PCManticore** [comment]
Hi @rogalski 

This is really nice, thank you! I will do a thorough review tomorrow. You can ignore the Travis failure for now, at least the one from PyPy ( I didn't have time to investigate it yet).

File: pylint/checkers/base.py [hunk 1]
```diff
@@ -1215,6 +1269,30 @@ def open(self):
...
+        naming_style = NAMING_STYLES[self.config.naming_style]
+
+        for name_type in naming_style.known_node_types():
+            regexps[name_type] = naming_style.get_regex(name_type)
+            overriden_value_name = "%s_rgx" % (name_type,)
+            overriden_value = getattr(self.config, overriden_value_name, None)
+            if overriden_value is not None:
+                regexps[name_type] = overriden_value
+
+            hints[name_type] = regexps[name_type].pattern
```
Comment (rogalski) [inline pylint/checkers/base.py]: If we define naming convention presets in code, maybe we can provide better hint than `pattern`. Maybe even hint could be a callable with single argument, where argument is a bad name provided by user. Now we can have non-static hint and suggest that `serviceURL` constant for snake_case should be renamed to `SERVICE_URL`.

File: pylint/checkers/base.py [hunk 2]
```diff
@@ -1123,30 +1167,31 @@ def visit_with(self, node):
...
 
 def _create_naming_options():
     name_options = []
-    for name_type, (rgx, human_readable_name) in six.iteritems(_NAME_TYPES):
+    for name_type in DEFAULT_NAMING_STYLE.known_node_types():
+        human_readable_name = HUMAN_READABLE_TYPES[name_type]
         name_type = name_type.replace('_', '-')
         name_options.append((
             '%s-rgx' % (name_type,),
-            {'default': rgx, 'type': 'regexp', 'metavar': '<regexp>',
```
Comment (PCManticore) [inline pylint/checkers/base.py]: Why do we have to change the defaults for these?

File: pylint/checkers/base.py (hunk 1 above)
Comment (PCManticore) [inline pylint/checkers/base.py]: Yeah, that would be interesting. So basically instead of 

```Invalid name service``

we could have

``Invalid name service, use SERVICE instead``?

Comment (rogalski) [inline pylint/checkers/base.py]: Yes, something like that. Just another step of making Pylint more user friendly (another feature similar to #1035 - don't tell me it's wrong, tell me how to fix it).

I'm aware that in general case it's quite hard and will be some kind of heuristic algorithm , but at the end, we can run heuristic result against predefined regex, and if it won't match just do not include suggestion out of message.

Sounds much better than these static hints that are available currently.

File: pylint/checkers/base.py (hunk 2 above)
Comment (rogalski) [inline pylint/checkers/base.py]: Because these defaults would override every regular expressions taken from preset. Without removing these defaults, naming-style has no effect at all (condition in line 1285 is always True) . 

Now, when `--generate-rcfile` is used, naming-style returns user_friendly name, and all of custom regular expressions are included as placeholders. Seems fair enough for for both normal users and power users.

Comment (PCManticore) [inline pylint/checkers/base.py]: But this is different than what the linked issue suggested. We should have naming-styles separated of the regular expressions, one could take precendence over the other though. For instance, if I want to customize how my variables should look like, with a custom pattern, how would I do it now?

Comment (rogalski) [inline pylint/checkers/base.py]:
```
naming-style=snake_case
variable-rgx=MY_NON_STANDARD_REGEX
```

All of naming styles are validated based on `snake_case` template, variables are validates based on MY_NON_STANDARD_REGEX.

**Comment by rogalski** [comment]
Oh, I re-read both issues and I guess I misunderstood it previously.
It suggested keeping individual naming styles for each type (module, variable, const etc.)
Here, there is one global naming style for project.

I think my approach is slightly better - how many projects reasonably require to have two naming conventions is same Python file?

**Comment by PCManticore** [comment]
That is not okay, we need to keep the same behavior for all the individual naming styles. For instance, you can have one naming conventions for classes and another one for methods, in case you are working with frameworks where they used camelCase convention previously (unittest, some packages from stdlib as well).

Ideally this patch should add just the naming style conventions, which can be used in turn of regexes when users want to, but it should not replace regexes per se. I am still looking into it and trying it right now.

**Comment by rogalski** [comment]
OK, thanks for review. I can give a shot for this ideal scenario, it should not be a big deal.

> _I am still looking into it and trying it right now.

Can you elaborate? I'll stop working on it if you write exact same functionality _right now_.

**Comment by PCManticore** [comment]
No, no, sorry for being confusing. I'm looking into this pull request, testing it. I'm not planning to write anything. :-)

**Comment by rogalski** [comment]
@PCManticore I want to clean up on my dangling PR - is this usable in any form or should be just thrown away?

**Comment by PCManticore** [comment]
I think this is a good start, but haven't been able to follow up with a more detailed analysis on how this change should be. Let's leave opened for now and as soon as I make some time, I will come up with a proper review and change plan. If you close it, there's a high chance of me forgetting about this if time passes, so being opened makes me aware of the proposed change.

**Comment by PCManticore** [comment]
@rogalski I managed to review this a bit today. It looks nice! What I think it needs now is a release note in What's new and maybe a bit some more tests to make sure we are covering all the known cases. Feel free to do this at your own pace, since we will probably merge it after 1.7.0.

**Comment by rogalski** [comment]
Why 'xxx-naming-hint' is _config value_ to `pylint`? I'd expect that it's autogenerated message based on other config values (those that actually affects how `pylint` checks naming convention).

**Comment by rogalski** [comment]
@PCManticore let me know what you think 660b2d9d1bf063f70119729d235f26f788955fd4 and 184b58a2d5bc392886b2eaa5bb6cf2e81c105b38. I think both `xxx-naming-style` and `include-naming-hint` may be removed with this new model of predefined name styles.

File: pylint/checkers/base.py [hunk 3]
```diff
@@ -1316,19 +1427,13 @@ def _find_name_group(self, node_type):
...
     def _raise_name_warning(self, node, node_type, name, confidence):
-        type_label = _NAME_TYPES[node_type][1]
-
-        hint_name = node_type.replace('_', '-') + '-name-hint'
-        hint_part = '%r template' % hint_name
-
-        if self.config.include_naming_hint:
-            hint_rgx = getattr(self.config, hint_name.replace('-', '_'))
-            hint_part += ' (hint: %r)' % hint_rgx
+        type_label = HUMAN_READABLE_TYPES[node_type]
```
Comment (PCManticore) [inline pylint/checkers/base.py]: I would still keep the naming hint flag (``include-naming-hint``) but only for naming styles, since the pattern already covers this. I'm thinking this would be a bit more helpful since there is no other way to see what we decided for a style to contain other than looking over the code (and the values won't be in the configuration either), so ``--include-naming-hint`` would be an user's only chance of seeing what a style value represents.

File: pylint/checkers/base.py [hunk 4]
```diff
@@ -1218,6 +1303,32 @@ def open(self):
...
+        regexps = {}
+        hints = {}
+
+        for name_type in KNOWN_NODE_TYPES:
+            naming_style_option_name = "%s_naming_style" % (name_type,)
+            naming_style_name = getattr(self.config, naming_style_option_name)
+
+            regexps[name_type] = NAMING_STYLES[naming_style_name].get_regex(name_type)
+
+            custom_regex_setting_name = "%s_rgx" % (name_type, )
```
Comment (PCManticore) [inline pylint/checkers/base.py]: There is a problem with picking the rgx if it exists, because currently we will always pick regexes, due to already having the ``-rgx`` hints pre-generated in config files. One idea would be to rename the options to something else, so for ``attr-rgx`` for instance, we should have ``attr-pattern`` or something along these lines. Could you think of another approach of how we should handle this?

File: pylint/checkers/base.py (hunk 3 above)
Comment (rogalski) [inline pylint/checkers/base.py]: Have you seen emitted message from current top of this PR? 

For out of the box settings:

    Function name "BADFUNCTION_name" doesn't conform to snake_case naming style_

For user defined settings:

    _Class name "classb" doesn't conform to '(?:(?P<UP>[A-Z]+)|(?P<down>[a-z]+))$' pattern_

I personally like them a lot, and to reintroduce include-naming-hint, I'd effectively replace this verbose and precise definitions with something arbitrary like _doesn't conform to selected naming style_ and _doesn't conform to custom user-defined regular expression_. Personally, I strongly prefer current form.

Include-naming-hint seems like a workaround for older messages where we emitted "Invalid class name 'classb'". With this new approach, where we provide both predefined and advanced mode for users, I don't see much value in keeping it.

File: pylint/checkers/base.py (hunk 4 above)
Comment (rogalski) [inline pylint/checkers/base.py]: > currently we will always pick regexes, due to already having the -rgx hints pre-generated in config files

Yes, and I think it's fine, and should be kept this way. By doing it we won't introduce any regressions to current clients. On the same time, new users will happily use this basic functionality.

We simply need reasonable documentation of new capabilities, both in changelog / what's new and in "naming convention" section in docs.

File: pylint/checkers/base.py (hunk 3 above)
Comment (PCManticore) [inline pylint/checkers/base.py]: Yes, I saw how it looks. Even if we show ``snake_case`` in the message, how an user will see what that means exactly without looking over the actual code, since these values are hardcoded in the code itself? I disagree about it being a workaround, especially since it's going to work for naming styles only

File: pylint/checkers/base.py (hunk 4 above)
Comment (PCManticore) [inline pylint/checkers/base.py]: Alright lets add documentation about this behaviour then

**Comment by PCManticore** [comment]
@rogalski This looks super nice, thank you! I just noticed you still have a couple of TODOs left in the tests, do you plan to fix those? Otherwise let's get this merged :)

**Comment by rogalski** [comment]
Hey Claudiu,
I'd take a last look at it somewhere in next week, clean that up (at least remove those todo comments) and merge it afterwards.

Thanks for approval!

**Comment by rogalski** [comment]
Merged, thanks for all valuable input!
